using_restricted_array.rb

```diff
@@ -7,41 +7,100 @@
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
 def length(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  length = 0
+  while array[i] != nil
+    length += 1
+    i += 1
+  end
+  return length
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  while array[i] != nil
+    puts array[i]
+    i += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  while i < length
+    if array[i] == value_to_find
+      return true
+    else
+      i += 1
+    end
+  end
+    return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  largest = array[0]
+
+  i = 0
+  while array[i] != nil
+    if array[i] > largest
+      largest = array[i]
+    else
+    end
+    i += 1
+  end
+  return largest
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  smallest = array[0]
+
+  i = 0
+  while array[i] != nil
+    if array[i] < smallest
+      smallest = array[i]
+    else
+    end
+    i += 1
+  end
+  return smallest
 end
 
 # Reverses the values in the integer array in place
 def reverse(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0 #first index
+  j = length - 1 #last index
+
+  while i < j
+    array[0] = array[length - 1]
```

Comment:
There's a bug here. You are missing the use of a temporary variable to swap. And in this case, you are just copying the last value over the to first value. (If array[0] is 3, and array[length-1] is 9. Line 80 copies 9 to be value of array[0]. And line 81 makes array[length - 1] the same value it already is, i.e. 9. Then it keeps copying 9 over and over again to array[0] and array[length -1]. This happens as long as i is less than j i.e. roughly length/2 number of times.

So, we need a temporary variable and leverage that, as well as leverage i and j.
` # array[i] is 3 and array[j] is 9, then
`temp = array[i] # make temp hold the value 3` 
`array[i] = array[j] # copy 9 to be the value of array[i]`
` array[j] = temp # copy 3 to be the value of array[j]`
`# array[i[ is now 9, array[j] is now 3`